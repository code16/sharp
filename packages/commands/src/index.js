
export default function(Vue, { store, router }) {
    
}

export * from './components';
export { commands as commandsModule } from './store';