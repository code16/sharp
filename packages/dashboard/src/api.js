import { api } from 'sharp';

export function getDashboard({ dashboardKey, filters }) {
    return api.get(`dashboard/${dashboardKey}`, {
        params: {
            ...filters,
        },
    }).then(response => response.data);
}

export function postDashboardCommand({ dashboardKey, commandKey, query, data }) {
    return api.post(`dashboard/${dashboardKey}/command/${commandKey}`, {
        query,
        data,
    }, { responseType: 'blob' });
}

export function getDashboardCommandFormData({ dashboardKey, commandKey, query }) {
    return api.get(`dashboard/${dashboardKey}/command/${commandKey}/data`, {
        params: {
            ...query,
        },
    }).then(response => response.data.data);
}