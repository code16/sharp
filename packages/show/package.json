{
    "name": "sharp-show",
    "private": true,
    "version": "1.0.0",
    "main": "src/index.js",
    "dependencies": {
        "sharp-commands": "file:../commands",
        "sharp-entity-list": "file:../entity-list",
        "sharp-filters": "file:../filters",
        "sharp-ui": "file:../ui"
    }
}
