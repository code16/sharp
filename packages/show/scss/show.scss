
@import 'page';
@import 'components/field';
@import 'components/EntityList';
@import 'components/Picture';
@import 'components/Text';
@import 'components/File';
@import 'components/List';