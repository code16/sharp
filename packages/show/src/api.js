import { api, apiUrl } from 'sharp';

export function getShowView({ entityKey, instanceId }) {
    return api.get(`show/${entityKey}/${instanceId || ''}`)
        .then(response => response.data);
}

export function postShowCommand({ entityKey, instanceId, commandKey, data }) {
    return api.post(`show/${entityKey}/command/${commandKey}/${instanceId ||Â ''}`, {
        data,
    }, { responseType: 'blob' });
}

export function getShowCommandFormData({ entityKey, instanceId, commandKey }) {
    return api.get(`show/${entityKey}/command/${commandKey}${instanceId ? `/${instanceId}` : ''}/data`)
        .then(response => response.data.data);
}

export function postShowState({ entityKey, instanceId, value }) {
    return api.post(`show/${entityKey}/state/${instanceId || ''}`, { value })
        .then(response => response.data);
}

export function downloadFileUrl({ entityKey, instanceId, fieldKey, fileName }) {
    return apiUrl(`show/download/${fieldKey}/${entityKey}/${instanceId}`, {
        params: {
            fileName,
        },
    })
}