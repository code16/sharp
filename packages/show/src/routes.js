import ShowPage from './components/pages/ShowPage.vue';

export default [
    {
        name: 'show',
        path: '/show/:entityKey/:instanceId?',
        component: ShowPage,
    },
]