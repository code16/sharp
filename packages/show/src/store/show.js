import { filtersModule as filters } from 'sharp-filters';
import { getShowView, postShowCommand, getShowCommandFormData, postShowState } from "../api";

const SET_ENTITY_KEY = 'SET_ENTITY_KEY';
const SET_INSTANCE_ID = 'SET_INSTANCE_ID';
const SET_SHOW = 'SET_SHOW_VIEW';


export default {
    namespaced: true,
    modules: {
        filters,
        'entity-lists': {
            namespaced: true,
        }
    },

    state: {
        entityKey: null,
        instanceId: null,
        show: null,
    },

    mutations: {
        [SET_SHOW](state, show) {
            state.show = show;
        },
        [SET_ENTITY_KEY](state, entityKey) {
            state.entityKey = entityKey;
        },
        [SET_INSTANCE_ID](state, instanceId) {
            state.instanceId = instanceId;
        },
    },

    getters: {
        entityKey(state) {
            return state.entityKey;
        },
        instanceId(state) {
            return state.instanceId;
        },
        config(state) {
            return state.show.config;
        },
        fields(state) {
            return state.show.fields;
        },
        layout(state) {
            return state.show.layout;
        },
        data(state) {
            return state.show.data;
        },
        breadcrumb(state) {
            return state.show.breadcrumb;
        },
        authorizations(state) {
            return state.show.authorizations;
        },
        canEdit(state, getters) {
            return getters.authorizations.update;
        },
        canChangeState(state, getters) {
            return !!getters.config.state && getters.config.state.authorization;
        },
        authorizedCommands(state, getters) {
            const commands = getters.config.commands ||Â {};
            return (commands.instance || [])
                .map(group => group.filter(command => command.authorization));
        },
        instanceState(state, getters) {
            const stateOptions = getters.config.state;
            if(stateOptions) {
                const stateValue = getters.data[stateOptions.attribute];
                return stateOptions.values.find(item => item.value === stateValue);
            }
            return null;
        },
        stateValues(state, getters) {
            return getters.config.state
                ? getters.config.state.values
                : null;
        },
    },

    actions: {
        async get({ state, commit }) {
            const data = await getShowView({
                entityKey: state.entityKey,
                instanceId: state.instanceId,
            });
            commit(SET_SHOW, data);
        },
        postCommand({ state }, { command, data }) {
            return postShowCommand({
                entityKey: state.entityKey,
                instanceId: state.instanceId,
                commandKey: command.key,
                data,
            });
        },
        getCommandFormData({ state }, { command }) {
            return getShowCommandFormData({
                entityKey: state.entityKey,
                instanceId: state.instanceId,
                commandKey: command.key,
            });
        },
        postState({ state, getters }, value) {
            return postShowState({
                entityKey: state.entityKey,
                instanceId: state.instanceId,
                attribute: getters.config.state.attribute,
                value,
            });
        },
        setEntityKey({ commit }, entityKey) {
            commit(SET_ENTITY_KEY, entityKey);
        },
        setInstanceId({ commit }, instanceId) {
            commit(SET_INSTANCE_ID, instanceId);
        },
    },
}