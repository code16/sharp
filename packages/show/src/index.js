import show from './store/show';
import routes from './routes';
import Field from './components/Field';

export default function (Vue, { router, store }) {
    router.addRoutes(routes);
    store.registerModule('show', show);
    Vue.component('ShowField', Field);
}