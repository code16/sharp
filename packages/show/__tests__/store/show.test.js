import Vuex from 'vuex';
import * as api from '../../src/api';
import show from '../../src/store/show';
import { createLocalVue } from '@vue/test-utils';


describe('store show', () => {
    test('state match snapshot', () => {
        expect(show.state).toMatchSnapshot();
    });

    test('canChangeState', () => {
        expect(show.getters.canChangeState(null, {
            config: {},
        })).toBe(false);
        expect(show.getters.canChangeState(null, {
            config: {
                state: {
                    authorization: false,
                }
            },
        })).toBe(false);
        expect(show.getters.canChangeState(null, {
            config: {
                state: {
                    authorization: true,
                }
            },
        })).toBe(true);
    });

    test('authorizedCommands', () => {
        expect(show.getters.authorizedCommands(null, {
            config: {},
        })).toEqual([]);
        expect(show.getters.authorizedCommands(null, {
            config: {
                commands: {
                    instance: [
                        [{ authorization: true }], [{ authorization:false }]
                    ]
                }
            },
        })).toEqual([[{ authorization: true }], []]);
    });

    test('instanceState', () => {
        expect(show.getters.instanceState(null, {
            config: {},
        })).toEqual(null);
        expect(show.getters.instanceState(null, {
            config: {
                state: {
                    attribute: 'state',
                    values: [{
                        value: 'currentState'
                    }]
                }
            },
            data: {
                state: 'currentState',
            }
        })).toEqual({ value:'currentState' });
    });

    test('stateValues', () => {
        expect(show.getters.stateValues(null, {
            config: {},
        })).toEqual(null);
        expect(show.getters.stateValues(null, {
            config: {
                state: {
                    values: [{
                        value: 'currentState'
                    }]
                }
            },
        })).toEqual([{ value:'currentState' }]);
    });

    test('get integration', async () => {
        const localVue = createLocalVue();
        localVue.use(Vuex);
        const store = new Vuex.Store(show);

        jest.spyOn(api, 'getShowView').mockImplementation(() => Promise.resolve({
            config: Symbol('config'),
            fields: Symbol('fields'),
            data: Symbol('data'),
            layout: Symbol('layout'),
            breadcrumb: Symbol('breadcrumb'),
            authorizations: Symbol('authorizations')
        }));

        await store.dispatch('setEntityKey', 'entityKey');
        await store.dispatch('setInstanceId', 'instanceId');
        await store.dispatch('get');

        expect(api.getShowView).toHaveBeenCalledWith({
            entityKey: 'entityKey',
            instanceId: 'instanceId',
        });

        const { config, fields, data, layout, breadcrumb, authorizations, } = store.getters;
        expect({
            config,
            fields,
            data,
            layout,
            breadcrumb,
            authorizations,
        }).toMatchSnapshot();
    });
});