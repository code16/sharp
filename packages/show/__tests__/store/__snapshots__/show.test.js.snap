// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store show get integration 1`] = `
Object {
  "authorizations": Symbol(authorizations),
  "breadcrumb": Symbol(breadcrumb),
  "config": Symbol(config),
  "data": Symbol(data),
  "fields": Symbol(fields),
  "layout": Symbol(layout),
}
`;

exports[`store show state match snapshot 1`] = `
Object {
  "entityKey": null,
  "instanceId": null,
  "show": null,
}
`;
